src/Blazored.LocalStorage/LocalStorageService.cs

```diff
@@ -20,7 +20,16 @@ public Task SetItem(string key, object data)
             if (string.IsNullOrEmpty(key))
                 throw new ArgumentNullException(nameof(key));
 
-            return _jSRuntime.InvokeAsync<object>("Blazored.LocalStorage.SetItem", key, Json.Serialize(data));
+            RaiseOnChanging(key, data, out ChangingEventArgs e);
+
+            if (e.Cancel)
+                return Task.CompletedTask;
+
+            var jsResult = _jSRuntime.InvokeAsync<object>("Blazored.LocalStorage.SetItem", key, Json.Serialize(data));
+
+            RaiseOnChanged(key, e.OldValue, data);

```

Comment:
I'm a little worried that this might not work as expected?  Since `InvokeAsync` above is not being awaited, `RaiseOnChanged` will actually (potentially) fire before `SetItem` has completed.

Reply:
I think you're right

Reply:
Actually, I'm confused about the contract of `public Task SetItem(string key, object data)`.

* shouldn't it be `async`?
* shouldn't it return `Task<object>` instead of `Task`? Currently, it seems to return the `jsRuntime` return value as a `Task`?

Reply:
A method only needs to be `async` if it uses `await` -- this makes the compiler create the whole async state machine (and thus should be avoided if not using `await`).  Since the method isn't awaiting, it can just return the `Task` outright (and `Task<T>` is a subclass of `Task` which is why you can return the former as the latter)

It _could_ return `Task<object>` but the only reason we're passing a type to `Invoke` at all, is because there isn't an overload that doesn't take a generic type argument -- the JS function doesn't return a value anyway.

tl;dr you should be able to make the method `async` and change the code to:

```
await _jSRuntime.InvokeAsync<object>("Blazored.LocalStorage.SetItem", key, Json.Serialize(data));

RaiseOnChanged(key, e.OldValue, data);
```

And then not return anything from the method at all.

Reply:
Ah, saw your update.  Yeah, you can just discard the return value (`_`) entirely, and note that technically your final `return;` statement is now unnecessary.

Reply:
I basically did that. Check my later commits.